### ISSUE TITLE
possible bug in `position!`

### ISSUE THREAD

**Issue by orialb** [ITensor/ITensors.jl#86]
Hi,

It seems to me that `position!` is not shifting the orthogonality center properly, in the code below the last test fails. Am I misunderstanding the functionality of `position!` or missing something else here? 
I can try to find out the issue if it is indeed a bug. 
```julia
using ITensors,LinearAlgebra, Test
N = 20
sites = spinOneSites(N)
ampo = AutoMPO(sites)
for j=1:N-1
    add!(ampo,"Sz",j,"Sz",j+1)
    add!(ampo,0.5,"S+",j,"S-",j+1)
    add!(ampo,0.5,"S-",j,"S+",j+1)
end
H = toMPO(ampo)
psi0 = randomMPS(sites)
sweeps = Sweeps(1)
maxdim!(sweeps, 10)
cutoff!(sweeps, 1E-10)
energy, psi = dmrg(H,psi0, sweeps)

position!(psi,5)
@test inner(psi,psi) ≈ 1.
@test dot(psi[5],psi[5]) ≈ 1.
```

**Comment by orialb** [comment]
I looked at it a bit more, and it seems that the problem is actually in `replaceBond!` which does not update the MPS left/right orthogonality limit. 
With the definition of `replaceBond!` below, everything seems to be fine.  Does it make sense to handle this in that way? 
```julia
function replaceBond!(M::MPS,
                      b::Int,
                      phi::ITensor;
                      kwargs...)

  FU,FV = factorize(phi,inds(M[b]); which_factorization="automatic",
                        tags=tags(linkindex(M,b)), kwargs...)
  M[b]   = FU
  M[b+1] = FV

  dir::String = get(kwargs,:dir,"center")
  if dir=="fromright"
    M.llim_ = b-1
    M.rlim_ = b+1
  elseif dir=="fromleft"
    M.llim_ = b
    M.rlim_ = b+2
  end
end
```

**Comment by emstoudenmire** [comment]
Hi Ori, thanks for taking a look at this. I think there were numerous issues with some interplay between them. One that Matt identified today when I was talking with him was that setindex!(M::MPS,...) was not updated the .llim_ and .rlim_ internal variables of the MPS as it should be. I just fixed that on Katie's ksh/morempo pull request branch.

Also after a discussion with Matt, we decided the behavior of position! in the C++ version is too confusing, because we mostly use it in a non-truncating mode, but sometimes we use it to truncate. When not used properly it does a weird thing of truncating different amounts of bonds depending on the history of how the MPS was modified.

So instead we're currently thinking that position! will just never truncate, and we will instead create a new function truncate! that will, and will truncate globally (every MPS bond every time it is called).

**Comment by orialb** [comment]
Hi Miles, thanks for your reply.
 With the change you made to `setindex!` in https://github.com/ITensor/ITensors.jl/commit/b45c3826fd86f702b17fe7a432e94760873dd417, I think that replaceBond! will have the following behavior (say we are using it during the left pass of a dmrg sweep)
```julia
# current bond being updated=b, leftLim(M) = b, rightLim(M)=b+2
M[b] = FU   # leftLim(M)=b-1
M[b+1] = FV # rightLim(M) = b+2
```
while in reality we would like to have rightLim(M) = b+1 after the update because M[b+1] is in right-normalized form. Am I misinterpreting the meaning of llim_ and rlim_ here?
Or is the idea that the caller of `setindex!` is responsible for handling shifting the orthogonality limits if it has extra information about the tensor which is being set  (e.g. that it is in right-normalized form) ?

**Comment by emstoudenmire** [comment]
Hi Ori,
Good question - yes in code like that one is responsible for updating llim_ and rlim_ manually. That's the reason we provide helper functions like replaceBond! so that users don't have to change llim_ and rlim_ themselves, but instead call replaceBond! and that library code does it for them. Worst case, if a user does use MPS setindex! it will later lead to extra QR's being done by position! or similar, but that's not so much extra overhead in many cases.

Miles

**Comment by emstoudenmire** [comment]
I'm going to mark this fixed by pull #72
